[FIPS] Use xxhash64 to calculate buildpack paths

rkoster opened the pull request:
Related CC PR: https://github.com/cloudfoundry/cloud_controller_ng/pull/3562

Fixes: https://github.com/cloudfoundry/cloud_controller_ng/issues/3561

ebroberson reviewed (approved):
LGTM

rkoster commented:
@winkingturtle-vmw yes both versions would have to be bumped at the same time into cf-deployment (and TAS). So this poses some cross-working group coordination. Discussion about his is happening here: https://github.com/cloudfoundry/cloud_controller_ng/pull/3562#issuecomment-1861350627

I'm also open to adding fallback support where it would try sha256 paths first, but if these are not found it will try md5. This way cloud controller will be fine as long as buildpacklifecycle gets bumped first.

winkingturtle-vmw commented:
This change has to be both ways backward compatible. We simply can't cut a release of diego if cloud_controller can't understand SHA256. If we wanted a more standard adaption, BAL (buildpackapplifecycle) has to emit both MD5 and SHA256 and capi-release should have the logic to prefer SHA256 over MD5. If it picked MD5, it should emit a warning/deprecation message that it will be removed. In a subsequent capi-release, it will remove the logic and BAL can then safely stop emitting MD5 files.

philippthun commented:
I've tried to create a version of `buildpack_app_lifecycle` that supports both variants (through a symlink). A commit with the changes can be found [here](https://github.com/sap-contributions/buildpackapplifecycle/commit/75c79889a9ae87945a53ab4ea240826b5a4e9297). After deploying this version, I was able to push an app through a Cloud Controller sending MD5-style paths as well as a Cloud Controller that sends SHA256-style paths. For the first push the staging output showed:
```
Copy [/tmp/buildpacks/8e04325d659d6865d53315c2a03131d7/dependencies/3ede9dafecf667b7774805a9bf68d63c/node_20.6.0_linux_x64_cflinuxfs4_6a8e7ea7.tgz]
```
And for the second push the following was shown:
```
Copy [/tmp/buildpacks/c9572f5ebd2c5f9c5ae3a7b531d0b64672bda5b2bc0dafed3c81ef717f25cd5c/dependencies/3ede9dafecf667b7774805a9bf68d63c/node_20.6.0_linux_x64_cflinuxfs4_6a8e7ea7.tgz]
```

Please note that I do not fully understand all the workflows this `buildpack_app_lifecycle` is used for, haven't adapted the tests, but simply tried to "make it work". Please treat this as input for the ongoing discussion...

rkoster commented:
Updated the PR to use xxhash64 as discussed [here](https://github.com/cloudfoundry/cloud_controller_ng/pull/3562#issuecomment-1862781314). Also implemented md5 fallback taking inspiration from the example provided by @philippthun [here](https://github.com/cloudfoundry/buildpackapplifecycle/pull/65#issuecomment-1862768039).

winkingturtle-vmw commented:
Commenting on the https://github.com/cloudfoundry/cloud_controller_ng/pull/3562 instead.

ebroberson commented:
I agree with @winkingturtle-vmw [here](https://github.com/cloudfoundry/cloud_controller_ng/pull/3562#issuecomment-1866211025) that SHA256 would be preferred if it can be used.

rkoster commented:
I have 2 concerns when it comes to SHA256:
1. It would increase the length of the buildpack dir path, which could be seen as a breaking change (if paths would have been close to the os max file path).
2. There already are c[onversations going on about dropping SHA1 support](https://www.nist.gov/news-events/news/2022/12/nist-retires-sha-1-cryptographic-algorithm), so we will have to keep changing the buildpack hashing algorithm to keep up. Given the input to the hash functions was public / non-sensitive (buildpack names) there is no need to use a cryptographic hash, and no need to keep up with NIST.

ebroberson commented:
Is the change from MD5 related to FIPS compliance?  That's what the link chain seems to imply, but if there's no need for a cryptographically secure algorithm, why change it?  I'm assuming there might be some internal tooling that's making the staging process fail, but would like to understand more.

rkoster commented:
Yes its FIPS related. Ruby uses openssl for calculating an MD5 hash. On a FIPS stemcell MD5 has been disable, so openssl will error out.

ebroberson commented:
Ahh, that makes sense, then.  Sounds good to me!

winkingturtle-vmw commented:
@rkoster Have you been able to actually deploy these changes successfully with CC? I haven't been unable to actually verify if this works, but I have left some comment on this PR.

rkoster commented:
Yeah have deployed these changes myself and was able to successfully run the smoke tests against a CF using FIPS stemcells.

@winkingturtle-vmw where did you leave comments?

winkingturtle-vmw commented:
@rkoster the code comment should be reflected right above this comment. It's referring to using a different import path for the library
